aliases.nix: convert or remove aliases older than 2019-06-01

Artturin opened the pull request:
old non-throw aliases are converted into throws
and old throws are removed

./alias/maintainers/scripts/remove-old-aliases.py --file ./pkgs/top-level/aliases.nix --year 2019 --month 6

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done
Converted and removed with my script https://github.com/NixOS/nixpkgs/pull/152162


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Artturin commented on pkgs/top-level/aliases.nix:
```diff
@@ -155,24 +155,24 @@ mapAliases ({
-  clutter_gtk = clutter-gtk; # Added 2018-02-25
+  clawsMail = throw "'clawsMail' has been renamed to/replaced by 'claws-mail'"; # Converted to throw 2022-02-21
+  clutter_gtk = throw "'clutter_gtk' has been renamed to/replaced by 'clutter-gtk'"; # Converted to throw 2022-02-21
   cmakeWithQt4Gui = throw "cmakeWithQt4Gui has been removed in favor of cmakeWithGui (Qt 5)"; # Added 2021-05
```
Note to self: make the script handle dates without days

Artturin replied:
fixed

vcunat reviewed (commented):
LGTM, though it's not like I really read each line.

I know the `ocaml-ng` parts are old, but they seem to be removed directly, without passing through the `throw` phase.  And I can't see why; it doesn't match the commit description.

Artturin commented:
> the `ocaml-ng` parts are old, but they seem to be removed directly, without passing through the `throw` phase. And I can't see why; it doesn't match the commit description.

Forgot to turn them into throws, I'll convert them later but I'm not sure how to

If you know then feel free to push directly to the pr branch

Artturin commented:
i do not know when the old ocaml aliases should be removed so i wont remove them, i removed the `Added 2016-09-14` comment. @Zimmi48 has removed ocaml packages before so i'll leave those to him

Artturin committed:
aliases.nix: convert or remove aliases older than 2019-06-01

old non-throw aliases are converted into throws
and old throws are removed

./alias/maintainers/scripts/remove-old-aliases.py --file ./pkgs/top-level/aliases.nix --year 2019 --month 6

Zimmi48 commented:
Yes, please do deprecate the OCaml aliases. All recent OCaml versions are only available through `ocaml-ng`, so there really is no reason that these aliases are kept.

Artturin committed:
aliases.nix: deprecate ocaml aliases

Artturin commented:
> Yes, please do deprecate the OCaml aliases. All recent OCaml versions are only available through `ocaml-ng`, so there really is no reason that these aliases are kept.

Done, please take a look.

```
  # TODO(ekleog): add ‘wasm’ alias to ‘ocamlPackages.wasm’ after 19.03
  # branch-off
```
should i remove this alias?

Zimmi48 commented on pkgs/top-level/aliases.nix:
```diff
@@ -1244,86 +1242,93 @@ mapAliases ({
+  ocamlPackages_4_03 = throw "'ocamlPackages_4_03' has been renamed to/replaced by 'ocaml-ng.ocamlPackages_4_03";
+  ocamlPackages_latest = throw "'ocamlPackages_latest' has been renamed to/replaced by 'ocaml-ng.ocamlPackages_latest";
+
+  ocaml_4_00_1 = throw "'ocamlPackages_4_00_1.ocaml' has been renamed to/replaced by 'ocaml-ng.ocaml_4_00_1.ocaml";
```
If I'm not mistaken this would rather be:

```suggestion
  ocaml_4_00_1 = throw "'ocamlPackages_4_00_1.ocaml' has been renamed to/replaced by 'ocaml-ng.ocamlPackages_4_00_1.ocaml";
```
And similarly on the other lines.

Zimmi48 commented:
> ```
>   # TODO(ekleog): add ‘wasm’ alias to ‘ocamlPackages.wasm’ after 19.03
>   # branch-off
> ```
> should i remove this TODO?

No idea what this is about. Let's ask @Ekleog

Artturin commented:
forgot to add dates to the ocaml throws

teto commented:
just rebased on unstable and had to rename it. I distinguish two types of aliases, the ones to handle renaming (typically a '_' that becomes a '-') for which it is LGTM to hard deprecate but some other aliases are cheap links to help users find software. For instance I've add `tshark` because that's the traditional (package) name for instance on ubuntu https://packages.ubuntu.com/search?keywords=tshark&searchon=names&suite=impish&section=all . The alias made tshark more discoverable so I am against its deletion. Not a hill I am willing to die on though.

vcunat commented:
Permanent aliases (or non-deprecated) should not be in `aliases.nix` but just among the normal attributes, I think.

teto commented:
I see there is a `### Deprecated aliases - for backward compatibility` mention that wasn't there when I added the `tshark` alias. I will do that and maybe even add your comment after `### Deprecated aliases - for backward compatibility`

vcunat commented:
One detail I recall is that OfBorg CI is running with `nixpkgs.config.allowAliases = false`

Profpatsch commented:
Thanks for causing half an hour of unnecessary work for me :+1:

vcunat commented:
Any better suggestion than keeping all aliases forever?

Profpatsch commented:
Keeping aliases forever, there is no hurt.

sternenseemann commented:
> Any better suggestion than keeping all aliases forever?

I think at the very least aliases should show a warning when used, just like renamed module options. This gives users a chance to update their use, as opposed to randomly breaking. There is no real user facing way to get notified about something becoming an alias.

vcunat commented:
> The problem with that is that it becomes a blessed name as well (and shows up in search etc.) which is maybe not a good idea always?

Maybe, but I can't see sufficient motivation to add more categories to the two that we have (blessed and deprecated).

Artturin commented:
@Profpatsch fyi you said in your commit message
> I replaced the packages with the ones the error messages mentioned, I
have not checked whether they are actually a no-op replacement.

they're no-op because they're the same packages you were already using


@sternenseemann
> I think at the very least aliases should show a warning when used, just like renamed module options.

this would be good and probably should have been implemented when the alias system was added

@teto 
https://github.com/NixOS/nixpkgs/pull/162511

sternenseemann commented:
We should at least implement it before we do the next removal round to avoid such a situation again.

vcunat commented:
IIRC usually such attempts failed on the fact that the warnings were too difficult to avoid in cases like `nix-env -qa`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/11
